core/src/main/java/org/mule/runtime/core/api/registry/AbstractServiceRegistry.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.api.registry;
+
+import static com.google.common.collect.ImmutableList.copyOf;
+import static java.lang.String.format;
+import static org.apache.commons.lang3.StringUtils.join;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+/**
+ * Base implementation of {@link ServiceRegistry} that enforce common behaviour for all implementations.
+ */
+public abstract class AbstractServiceRegistry implements ServiceRegistry {
+
+  /**
+   * If {@code classLoader} is {@code null}, then the current {@link Thread#getContextClassLoader()} will be used {@inheritDoc}
+   */
+  @Override
+  public final <T> Collection<T> lookupProviders(Class<T> providerClass, ClassLoader classLoader) {
+    if (classLoader == null) {
+      classLoader = Thread.currentThread().getContextClassLoader();
+    }
+
+    return copyOf(doLookupProviders(providerClass, classLoader));
+  }
+
+  /**
+   * Retrieves a collection of providers. This method will be reused to actually implement {@code {@link #lookupProviders(Class)},
+   * and {@code {@link #lookupProvider(Class)}}}.
+   *
+   * @param providerClass a <code>Class</code>object indicating the class or interface of the service providers being detected.
+   * @param loader the class loader to be used to load provider/configuration files and instantiate provider instances. If
+   *        {@code null}, it will be up to the implementation to choose a {@link ClassLoader}
+   * @return A {@link Collection} that yields provider objects for the given service, in some arbitrary order.
+   */
+  protected abstract <T> Collection<T> doLookupProviders(Class<T> providerClass, ClassLoader loader);
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public final <T> Collection<T> lookupProviders(Class<T> providerClass) {
+    return copyOf(ServiceLoader.load(providerClass).iterator());
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public final <T> T lookupProvider(Class<T> providerClass) {
+    Collection<T> providers = lookupProviders(providerClass, Thread.currentThread().getContextClassLoader());
```

Comment:
As with lookupProviders, there's both options, passing the classloader or not, in which case the current context class loader is used. The other method was not implemented since it's not required now.